### PR TITLE
fix: keep video frames of the same media in a single subset during assignment

### LINKED ISSUE #7619: Subset assignment splits frames of the same video across train/val/test, inflating evaluation metrics
**Describe the bug**

First of all — thank you for open-sourcing Geti; the API-first workflow has been a joy to drive end to end. While validating models on held-out footage we ran into something we believe is worth your attention.

`SubsetAssigner` (`application/backend/app/services/subset_assignment/assigner.py`) stratifies dataset items by label only — `DatasetItemWithLabels` carries no media identity. When a project's annotations are video frames, frames of the *same video* are distributed freely across TRAINING, VALIDATION and TESTING.

Since temporally adjacent frames of one video are near-duplicates, the test subset effectively contains copies of the training data, and evaluation metrics come out inflated — with no indication to the user. We verified the assigner is unchanged at `develop` @ `e75ff15f1`.

**Steps to Reproduce**

1. Create a detection project and upload a single video; annotate ~30 frames spread across it (one label).
2. Train any detection architecture (we tried YOLOX-Tiny and RF-DETR-M, on two different datasets).
3. Observe test metrics: we consistently got **mAP@0.5 = 1.0**.
4. Run the trained model on held-out footage of the same object (different scene/lighting): we measured as low as **48% frame-level recall** — while the in-app metric reads 1.0 with no caveat.

**Environment:**

- Geti app / library version: 3.1.0 (also verified code path unchanged on `develop` @ e75ff15f1)
- Installation mode: Docker (pre-built `ghcr.io/open-edge-platform/geti-cuda`)
- Hardware spec: RTX 5090 (32 GB), 64-thread CPU
- OS: Ubuntu 22.04 host
- GPU drivers: CUDA driver 580.178.04

**Suggested direction (happy to contribute)**

The schema already links frames to their parent video (`media.video_id`), so we prototyped:
1. Group-aware assignment — frames of the same video stay within one subset (group-level iterative stratification) whenever ≥3 distinct media groups exist;
2. A logged warning when leakage is unavoidable (e.g. all annotations come from a single video — arguably the most common beginner flow), noting that metrics likely overestimate real-world performance.

We have this working with unit + integration tests passing and would be glad to open a PR — equally happy to adapt to whatever direction the maintainers prefer. Thanks for taking a look!

### PR DESCRIPTION
**Pull request by shashb27**
### Summary

Fixes an evaluation-integrity gap for video-annotated projects: `SubsetAssigner` stratified dataset items by label only, so annotated frames of the *same video* could land in TRAINING, VALIDATION and TESTING simultaneously. Temporally adjacent frames are near-duplicates, which inflates evaluation metrics without any indication to the user (we observed test mAP@0.5 = 1.0 alongside 48% frame-level recall on genuinely held-out footage).

Resolves #7619.

Changes:
- `DatasetItemWithLabels` gains an optional `group_id` — the parent video id for video frames, `None` for images (which keep the existing per-item behavior).
- `SubsetAssigner` stratifies at the media-group level when ≥ 3 distinct groups exist, assigning whole groups to a single subset; group labels are the union of member-item labels.
- When grouping is impossible (e.g. all annotations come from a single video), it falls back to the previous per-item behavior and logs a warning that evaluation metrics may overestimate real-world performance.
- `DatasetItemRepository.list_unassigned_items` returns the parent video id; `SubsetService` wires it through.

Design notes, very open to feedback:
- `MIN_GROUPS_FOR_GROUP_AWARE_SPLIT = 3` (fewer groups than subsets cannot be split without leakage); happy to make this configurable.
- Group-level stratification approximates item-level ratios when group sizes are uneven — we felt leak-free evaluation is worth the small ratio drift, but a size-weighted approach could be added if preferred.
- The fallback-plus-warning path deliberately preserves current behavior for single-video projects so nothing breaks; surfacing the warning in the UI could be a follow-up.

### How to test

- `pytest tests/unit/services/subset_assignment/ tests/integration/services/subset_assignment/` — 34 tests pass (24 existing + 4 new group-behavior tests + 6 integration). New tests cover: groups never split across subsets; single-group fallback still assigns and warns; images unaffected; mixed videos + images.
- Manually reproduced the original issue on a Geti 3.1.0 project (single annotated video, 45 frames) and confirmed the warning path triggers; `ruff check` and `ruff format --check` pass on all touched files.

### Checklist

- [x] The PR title and description are clear and descriptive
- [x] I have manually tested the changes
- [x] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable — none of the user-facing docs describe subset internals; happy to add a note if you'd like one)

### CODE REVIEW

File: application/backend/app/repositories/dataset_item_repo.py [hunk 1]
```diff
@@ -209,16 +209,19 @@ def delete_label_from_items(self, label_id: str) -> None:
...
-    def list_unassigned_items(self) -> list[DatasetItemLabelDB]:
+    def list_unassigned_items(self) -> list[tuple[DatasetItemLabelDB, str | None]]:
+        """List label rows of unassigned dataset items, each paired with the parent
+        video id of the underlying media (None for images and standalone media)."""
         stmt = (
-            select(DatasetItemLabelDB)
-            .join(DatasetItemDB)
+            select(DatasetItemLabelDB, MediaDB.video_id)
+            .join(DatasetItemDB, DatasetItemLabelDB.dataset_item_id == DatasetItemDB.id)
+            .join(MediaDB, MediaDB.id == DatasetItemDB.id)
```
[the remark this answers is not in the document]
Comment (shashb27) [reply application/backend/app/repositories/dataset_item_repo.py:218]: Great catch, thank you — you're right that this only saw UNASSIGNED items, so incremental runs could split a video across subsets. Fixed in 9dbede88f: a new `list_assigned_group_subsets` query surfaces each video's already-assigned frames, `SubsetService.get_pinned_group_subsets` reduces that to a per-video subset (majority wins, deterministic tie-break), and the assigner now pins new frames of an established video to that subset before stratifying the rest. Covered by new unit tests and an integration test against the real schema.

File: application/backend/app/services/subset_assignment/assigner.py [hunk 2]
```diff
@@ -67,37 +78,93 @@ def assign(
                 assignments.append(SubsetAssignment(item_id=items[idx].item_id, subset=subsets[idx]))
             return assignments
 
-        label_matrix = self._mlb.fit_transform([item.labels for item in items])
+        items_by_group = self._group_items(items)
+        group_keys = list(items_by_group.keys())
+        has_multi_item_groups = any(len(indices) > 1 for indices in items_by_group.values())
+        use_groups = has_multi_item_groups and len(group_keys) >= MIN_GROUPS_FOR_GROUP_AWARE_SPLIT
```
[the remark this answers is not in the document]
Comment (shashb27) [reply application/backend/app/services/subset_assignment/assigner.py:97]: Thanks — agreed, the size guard bypassed the group logic entirely. Fixed in 9dbede88f: the fewer-than-three-items path now assigns media *groups* sequentially instead of individual items, so a small incremental batch of frames from one video lands in a single subset (unrelated single items keep the previous TRAINING→VALIDATION ordering for backward compatibility). Pins are also applied before the guard, so frames of an established video join its existing subset even in tiny batches. Both cases have dedicated unit tests.
